src/index.js

```diff
@@ -0,0 +1,264 @@
+ */
+class Client extends events.EventEmitter {
+  constructor({username, password, hostname, connectionString, recycleInterval, retirementDelay}) {
+    super();
+
+    if (connectionString) {
+      assert(!username, 'Can\'t use `username` along with `connectionString`');
+      assert(!password, 'Can\'t use `password` along with `connectionString`');
+      assert(!hostname, 'Can\'t use `hostname` along with `connectionString`');
+      this.connectionString = connectionString;
+    } else {
+      connectionString = buildConnectionString({username, password, hostname});
+    }
+
+    this.recycleInterval = recycleInterval || 3600 * 1000;
+    this.retirementDelay = retirementDelay || 30 * 1000;
+    this.running = false;
+    this.connections = [];
+    this.connectionCounter = 0;
+
+    this.id = ++clientCounter;
+    this.debug = debug(`taskcluster-lib-pulse:client:${this.id}`);
+  }
+
+  /**
+   * Start connecting and stay connected until stop() is called
+   */
+  start() {
+    assert(!this.running, 'Already running');
+    this.debug('starting');
+    this.running = true;
+    this.recycle();
+
+    this._interval = setInterval(
+      () => this.recycle(),
+      this.recycleInterval);
+  }
+
+  async stop() {
+    assert(this.running, 'Not running');
+    this.debug('stopping');
+    this.running = false;
+
+    clearInterval(this._interval);
+    this._interval = null;
+
+    this.recycle();
+
+    // wait until all existing connections are finished
+    const unfinished = this.connections.filter(conn => conn.state !== 'finished');
+    if (unfinished.length > 0) {
+      await Promise.all(unfinished.map(
+        conn => new Promise(resolve => { conn.once('finished', resolve); })));
+    }
+  }
+
+  /**
+   * Create a new connection, retiring any existing connection.
+   */
+  recycle() {
+    this.debug('recycling');
+
+    if (this.connections.length) {
+      const currentConn = this.connections[0];
+      currentConn.retire();
+    }
+
+    if (this.running) {
+      const newConn = new Connection(this, ++this.connectionCounter);
+      newConn.once('connected', () => {
+        this.emit('connected', newConn);
+      });
+      newConn.once('finished', () => {
+        this.connections = this.connections.filter(conn => conn.id !== newConn.id);
+      });
+      this.connections.unshift(newConn);
+    }
+  }
+}
+
+exports.Client = Client;
+
+/**
+ * A single connection to a pulse server.  This is a thin wrapper around a raw
+ * AMQP connection, instrumented to inform the parent Client of failures
+ * and trigger a reconnection.  It is possible to have multiple Connection
+ * objects in the same process at the same time, while one is being "retired" but
+ * is lingering around to send ack's for any in-flight message handlers.
+ *
+ * The instance's `amqp` property is the amqp connection object.  In the event of any
+ * issues with the connection, call the instance's `failed` method.  This will initiate
```

Comment:
Why don't we detect by listening to the `error` event... calling failed seems like a bad idea...

Reply:
We do that, too.  This is more of a belt-and-suspenders approach, where if the user sees some issue on the connection (like a timeout) they can also flag it as a failure.